src/game/commands/templates.ts

```diff
@@ -0,0 +1,105 @@
+import { any, hasType, isExact, isPositive, Template } from "./predicate";
+import { emptyToken, TokenStream, TokenType } from "./types/Token";
+import { Err, err, isErr, isOk, ok, Result } from "./types/Result";
+
+// Types -----------------------------------------------------------------------
+export type MatchResult =
+  [TokenStream, Template]
+
+export type MatchError =
+  string
+
+const NO_MATCHES = "I couldn't find any commands that matched this sequence";
+const UNRECGONISED_INTIAL_TOKEN = "The first token in this sequence is not a " +
+  "command I recognise.";
+const TOO_MANY_ARGUMENTS = (command: string, got: string) =>
+  "While parsing " + command + " I received an extra " + got + ".";
+
+// This type alias just helps to slim down some of the function signatures
+// below.
+type R = Result<MatchError, MatchResult>
+
+// GameCommand templates -------------------------------------------------------
+const gameCommandTemplates: Template[] = [
+  [isExact("/help")],
+  [isExact("/help"), any([hasType(TokenType.GameCommand), hasType(TokenType.PlayerCommand)])],
+];
+
+// PlayerCommand templates -----------------------------------------------------
+const playerCommandTemplates: Template[] = [
+  [any([isExact("walk"), isExact("move")]), hasType(TokenType.Direction)],
+  [any([isExact("walk"), isExact("move")]), isPositive, hasType(TokenType.Direction)],
```

Comment:
Does your parser support `[hasType(Direction)]`? If not, support for just typing a direction has been lost. (Not sure if that is an issue, but might be worth thinking about)

Reply:
Yeah, support for those should be added to player commands not here. It will be added later.